packages/core/src/indexing/service.ts

```diff
@@ -1,103 +1,134 @@
+    endCheckpoint?: Checkpoint;
+    eventsProcessed?: number;
   };
 };
-type LogEventTask = { kind: "LOG"; event: LogEvent };
+
 type IndexingFunctionTask = SetupTask | LogEventTask;
 type IndexingFunctionQueue = Queue<IndexingFunctionTask>;
 
+// Note: this should move to a dynamic value, based on how many indexing function keys there are.
+const TASK_BATCH_SIZE = 1_000;
+
 export class IndexingService extends Emittery<IndexingEvents> {
   private common: Common;
   private indexingStore: IndexingStore;
   private syncGatewayService: SyncGateway;
   private sources: Source[];
-  private contexts: Record<
-    number,
-    {
-      client: Client;
-      network: { chainId: number; name: string };
-      contracts: Record<
-        string,
-        {
-          abi: Abi;
-          address?: Address | readonly Address[];
-          startBlock: number;
-          endBlock?: number;
-          maxBlockRange?: number;
-        }
-      >;
-    }
-  > = {};
+  private networks: Network[];
 
-  private schema?: Schema;
-  private db: Record<string, DatabaseModel<any>> = {};
+  private isPaused = false;
 
   private indexingFunctions?: IndexingFunctions;
+  private schema?: Schema;
+  private tableAccess?: TableAccess;
 
-  private eventProcessingMutex: Mutex;
-  private queue?: IndexingFunctionQueue;
+  queue?: IndexingFunctionQueue;
 
-  private eventsProcessedToCheckpoint: Checkpoint = zeroCheckpoint;
+  private getNetwork: (checkpoint: Checkpoint) => Context["network"] =
+    undefined!;
+  private getClient: (checkpoint: Checkpoint) => Context["client"] = undefined!;
+  private getDB: (checkpoint: Checkpoint) => Context["db"] = undefined!;
+  private getContracts: (checkpoint: Checkpoint) => Context["contracts"] =
+    undefined!;
 
-  private currentIndexingCheckpoint: Checkpoint = zeroCheckpoint;
-  private isPaused = false;
-  /** If true, failed tasks should not log errors or be retried. */
-  private isShuttingDown = false;
+  private indexingFunctionMap: Record<
+    /* Indexing function key: "{ContractName}:{EventName}" */
+    string,
+    {
+      contractName: string;
+      eventName: string;
+      /* Indexing function keys that write to tables that this indexing function key reads from. */
+      parents: string[];
+      /* True if this key is a parent of itself. */
+      selfReliance: boolean;
+      /* Sources that contribute to this indexing function. */
+      sources: Source[];
+      abiEvent: AbiEvent;
+      eventSelector: Hex;
+
+      /* Checkpoint of max completed task. */
+      checkpoint: Checkpoint;
+      /* Checkpoint of the most recent task loaded from db. */
+      maxTaskCheckpoint: Checkpoint;
+      /* Buffer of in memory tasks that haven't been enqueued yet. */
+      indexingFunctionTasks: LogEventTask[];
+      /* Mutex ensuring tasks are loaded one at a time. */
+      dbMutex: Mutex;
+      /** True if a task has been enqueued with itself as the most recent checkpoint. */
+      serialQueue: boolean;
```

Comment:
```
      /* Checkpoint of max completed task. */
      tasksProcessedToCheckpoint: Checkpoint;
      /* Checkpoint of the most recent task loaded from db. */
      tasksLoadedToCheckpoint: Checkpoint;
      /* Buffer of in memory tasks that haven't been enqueued yet. */
      loadedTasks: LogEventTask[];
      /* Mutex ensuring tasks are not loaded twice. */
      loadingMutex: Mutex;
      /** Lock to prevent tasks from running concurrently. */
      selfDependentLock: boolean;
```